qutip/core/data/convert.pyx

```diff
@@ -260,61 +262,159 @@ cdef class _to:
+        """
+        Return a data-layer type object given its name or the type itself.
+
+        Parameters
+        ----------
+        dtype : type, str
+            Either the name of a data-layer type or a type itself.
+
+        Returns
+        -------
+        type
+            A data-layer type.
+
+        Raises
+        ------
+        TypeError
+            If ``dtype`` is neither a string nor a type.
+        ValueError
+            If ``dtype`` is a name, but no data-layer type of that name is
+            registered, or if ``dtype`` is a type, but not a known data-layer
+            type.
+        """
+        if type(dtype) is type:
+            if dtype not in self.dtypes:
+                raise ValueError(
+                    "Type is not a data-layer type: " + repr(dtype))
+            return dtype
+        elif type(dtype) is str:
+            try:
+                return self._str2type[dtype]
+            except KeyError:
+                raise ValueError(
+                    "Type name is not known to the data-layer: " + repr(dtype)
+                    ) from None
+
+        raise TypeError(
+            "Invalid dtype is neither a type nor a type name: " + repr(dtype))
+
     @cython.boundscheck(False)
     @cython.wraparound(False)
     def __getitem__(self, arg):
-        if isinstance(arg, type):
+        if type(arg) is not tuple:
             arg = (arg,)
-        if not isinstance(arg, tuple) or not arg or len(arg) > 2:
+        if not arg or len(arg) > 2:
             raise KeyError(arg)
-        to_t = arg[0]
-        if to_t not in self.dtypes:
-            raise TypeError("to_type is not known: " + str(to_t))
+        to_t = self.parse(arg[0])
         if len(arg) == 1:
             converters = {
                 from_t: self._convert[to_t, from_t] for from_t in self.dtypes
             }
             return _partial_converter(converters, to_t)
-        from_t = arg[1]
-        if from_t not in self.dtypes:
-            raise TypeError("from_type is not known: " + str(from_t))
+        from_t = self.parse(arg[1])
         return self._convert[to_t, from_t]
 
     def __call__(self, to_type, data):
-        if not isinstance(to_type, type):
-            raise TypeError(repr(to_type) + " is not a type object")
-        if to_type not in self.dtypes:
-            raise ValueError("unknown output type: " + to_type.__name__)
-        from_type = type(data)
-        if from_type not in self.dtypes:
-            raise TypeError("unknown input type: " + from_type.__name__)
+        to_type = self.parse(to_type)
+        from_type = self.parse(type(data))
         if to_type == from_type:
             return data
         return self._convert[to_type, from_type](data)
 
 
 cdef class _create:
+    cdef readonly list _creators
+
     def __init__(self):
-        pass
+        self._creators = []
 
-    def add_creators(self, creators):
-        pass
+    def add_creator(self, condition, creator, priority=0):
```

Comment:
What was the reason for making it only take a single one at a time?  This way seems inconsistent with `data.to.add_conversions`, and it introduces a quadratic dependency (complexity goes to O(n^2 log n)).  Not that the complexity will matter, really, but I'm not sure I see the reason, and it isn't consistent with `data.to`.

Reply:
`add_conversions` force rebuilding the dispatch table. So it's worth doing at once. 
But here the time cost is insignificant and the function cleaner that way.
Doing it one at a time feels more natural, and with the dispatch table, we are forced to do it at once in `add_conversions`.

Reply:
I'm not sure about it feeling more natural (iterables are very Pythonic), and not certain how it would be significantly more complicated in a loop, but ok.  My concern is less about necessity and more about consistency.  I'd imagine that if someone wants to implement a new type, they probably want to add conversions from several different types of object (just like we do), so as a user I'd expect the function to take several at once like with `add_conversions`.  How easy/difficult it is for us internally is secondary to consistent UX for me.  (In a similar vein, I might even make the argument that `register_alias` should be `register_aliases`, and allow either a string or an iterable as its `alias` argument.)

Reply:
Yes, `register_aliases(([CSR', 'csr'], CSR)` feels natural. Reverting.